### PR TITLE
Feature export plugins

### PR DESCRIPTION
**Pull request by MstrPikachu**

### CODE REVIEW

**Comment by MstrPikachu** [comment]
My addition to CommandsFactory is a bit sus

File: src/main/java/org/hydev/mcpm/client/arguments/CommandsFactory.java [hunk 1]
```diff
@@ -31,14 +26,15 @@ private CommandsFactory() { }
      */
     public static List<CommandParser> baseParsers() {
         var echoController = new EchoController();
+        var exportPluginsController = new ExportPluginsController(null); // do I have to instantiate the entire thing ??
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/arguments/CommandsFactory.java]: Yep, you do 😰

File: src/main/java/org/hydev/mcpm/client/arguments/CommandsFactory.java [hunk 2]
```diff
@@ -31,14 +26,15 @@ private CommandsFactory() { }
...
 
         /*
          * Add general parsers to this list!
          * All versions of MCPM will have access to these parsers.
          * If you're not sure if your command is server-only, add it to this list!
          */
         return List.of(
-            new EchoParser(echoController)
+            new EchoParser(echoController), new ExportPluginsParser(exportPluginsController)
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/arguments/CommandsFactory.java]: I mean... if I'm being pedantic...

```suggestion
            new EchoParser(echoController),
            new ExportPluginsParser(exportPluginsController)
```

File: src/main/java/org/hydev/mcpm/client/arguments/parsers/ExportPluginsParser.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    @Override
+    public @Nullable Subparser configure(Subparsers parsers) {
+        var parser = parsers.addParser("export"); // name of command
+
+        parser.addArgument("outfile") // add optional output file
+                .type(OutputStream.class).dest("outfile"); // of type OutputStream
+        parser.addArgument("-c", "--cache")
+                .type(boolean.class).dest("cache");
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/arguments/parsers/ExportPluginsParser.java:30]: > `boolean.class`

This is pretty cool that it builds. I would expect `Boolean.class`.

File: src/main/java/org/hydev/mcpm/client/database/export/Export.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import org.hydev.mcpm.client.database.fetcher.ProgressBarFetcherListener;
+import org.hydev.mcpm.client.database.inputs.ExportPluginsInput;
+import org.hydev.mcpm.client.database.results.ExportPluginsResult;
+
+import java.io.PrintStream;
+
+/**
+ * An implementation of ExportPluginsBoundary that fetches from a database.
+ */
+public class Export implements ExportPluginsBoundary {
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/database/export/Export.java]: 1 Use Case = 1 Interactor

```suggestion
public class ExportInteractor implements ExportPluginsBoundary {
```

File: src/main/java/org/hydev/mcpm/client/database/export/Export.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+    /**
+     * Outputs the plugins on each line as its name and version separated by a space.
+     *
+     * @param input The output stream to write to
+     * @return whether the export was successful or otherwise.
+     */
+    @Override
+    public ExportPluginsResult export(ExportPluginsInput input) {
+        var database = fetcher.fetchDatabase(input.cache(), new ProgressBarFetcherListener());
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/database/export/Export.java]: Hmm... I think you should be looking at the plugins that are currently installed. You can use `interface PluginTracker` for that. DatabaseFetcher fetches all plugins in existence.

File: src/main/java/org/hydev/mcpm/client/arguments/parsers/ExportPluginsParser.java (hunk 3 above)
Comment (MstrPikachu) [reply src/main/java/org/hydev/mcpm/client/arguments/parsers/ExportPluginsParser.java:30]: surely it works...

File: src/main/java/org/hydev/mcpm/client/arguments/CommandsFactory.java (hunk 1 above)
Comment (1whatleytay) [reply src/main/java/org/hydev/mcpm/client/arguments/CommandsFactory.java]: I should mention that `CommandsFactory` is just a suggestion (like a default). Anyone can make their own `List<CommandParser>` and configure it the way they want. This is just the default "right" way.

**Review by hykilpikonna — APPROVED** [review]

**Review by 1whatleytay — APPROVED** [review]

**Pull request merged**